subsys/bluetooth/host/classic/hfp_hf.c

```diff
@@ -295,390 +397,3513 @@ int cind_resp(struct at_client *hf_at, struct net_buf *buf)
+	int err;
+
+	LOG_DBG("");
+
+	if (!hf) {
+		LOG_ERR("No HF connection found");
+		return -ENOTCONN;
+	}
+
+	call = get_dialing_call(hf);
+	if (call) {
+		LOG_ERR("There is a call in alerting or waiting");
+		return -EBUSY;
+	}
+
+	call = get_new_call(hf);
+	if (!call) {
+		LOG_ERR("Not available call object");
+		return -ENOMEM;
+	}
+
+	hf_call_state_update(call, BT_HFP_HF_CALL_STATE_OUTGOING);
+
+	err = hfp_hf_send_cmd(hf, NULL, atd_finish, "ATD>%s", location);
+	if (err < 0) {
+		LOG_ERR("Fail to start memory dialing on %p", hf);
+	}
+
+	return err;
+}
+
+static int bldn_finish(struct at_client *hf_at, enum at_result result,
+		   enum at_cme cme_err)
+{
+	struct bt_hfp_hf *hf = CONTAINER_OF(hf_at, struct bt_hfp_hf, at);
+	int err;
+	struct bt_hfp_hf_call *call;
+
+	LOG_DBG("AT+BLDN (result %d) on %p", result, hf);
+
+	if (result == AT_RESULT_CME_ERROR) {
+		err = bt_hfp_ag_get_cme_err(cme_err);
+	} else if (result == AT_RESULT_ERROR) {
+		err = -ENOTSUP;
+	} else {
+		err = 0;
+	}
+
+	call = get_call_with_state(hf, BT_HFP_HF_CALL_STATE_OUTGOING);
+	if (!call) {
+		return -EINVAL;
+	}
+
+	if (err) {
+		free_call(call);
+	}
+
+	if (bt_hf && bt_hf->dialing) {
+		bt_hf->dialing(hf, err);
+	}
+	return 0;
+}
+
+int bt_hfp_hf_redial(struct bt_hfp_hf *hf)
+{
+	struct bt_hfp_hf_call *call;
+	int err;
+
+	LOG_DBG("");
+
+	if (!hf) {
+		LOG_ERR("No HF connection found");
+		return -ENOTCONN;
+	}
+
+	call = get_dialing_call(hf);
+	if (call) {
+		LOG_ERR("There is a call in alerting or waiting");
+		return -EBUSY;
+	}
+
+	call = get_new_call(hf);
+	if (!call) {
+		LOG_ERR("Not available call object");
+		return -ENOMEM;
+	}
+
+	hf_call_state_update(call, BT_HFP_HF_CALL_STATE_OUTGOING);
+
+	err = hfp_hf_send_cmd(hf, NULL, bldn_finish, "AT+BLDN");
+	if (err < 0) {
+		LOG_ERR("Fail to start memory dialing on %p", hf);
+	}
+
+	return err;
+}
+
+#if defined(CONFIG_BT_HFP_HF_CODEC_NEG)
+static int bcc_finish(struct at_client *hf_at, enum at_result result,
+		   enum at_cme cme_err)
+{
+	struct bt_hfp_hf *hf = CONTAINER_OF(hf_at, struct bt_hfp_hf, at);
+
+	LOG_DBG("BCC (result %d) on %p", result, hf);
+
+	return 0;
+}
+#endif /* CONFIG_BT_HFP_HF_CODEC_NEG */
+
+int bt_hfp_hf_audio_connect(struct bt_hfp_hf *hf)
+{
+#if defined(CONFIG_BT_HFP_HF_CODEC_NEG)
+	int err;
+
+	LOG_DBG("");
+
+	if (!hf) {
+		LOG_ERR("No HF connection found");
+		return -ENOTCONN;
+	}
+
+	if (hf->chan.sco) {
+		LOG_ERR("Audio conenction has been connected");
+		return -ECONNREFUSED;
+	}
+
+	err = hfp_hf_send_cmd(hf, NULL, bcc_finish, "AT+BCC");
+	if (err < 0) {
+		LOG_ERR("Fail to setup audio connection on %p", hf);
+	}
+
+	return err;
+#else
+	return -ENOTSUP;
+#endif /* CONFIG_BT_HFP_HF_CODEC_NEG */
+}
+
+#if defined(CONFIG_BT_HFP_HF_CODEC_NEG)
+static int bcs_finish(struct at_client *hf_at, enum at_result result,
+		   enum at_cme cme_err)
+{
+	struct bt_hfp_hf *hf = CONTAINER_OF(hf_at, struct bt_hfp_hf, at);
+
+	LOG_DBG("BCC (result %d) on %p", result, hf);
```

Comment:
log message not right,  it suppose  "BCS" not "BCC"

Reply:
updated